[APIC-305] Require sql_type for every column when table_columns is provided

izzy-zelichenko opened the pull request:
[APIC-305](https://civisanalytics.atlassian.net/browse/APIC-305)

Ensure data types are detected when the destination table does not exist or it is being dropped and table_columns are provided but do not contain a sql_type for every column.

uttercm commented on civis/io/_tables.py:
```diff
@@ -1029,8 +1029,16 @@ def civis_file_to_table(file_id, database, table, client=None,
     # Use Preprocess endpoint to get the table columns as needed
     # and perform necessary file cleaning
-    need_table_columns = ((not table_exists or existing_table_rows == 'drop')
+    table_columns_have_sql = True
```
nit: this seems a bit wordy maybe something like `table_columns_valid`, idk, thoughts?

izzy-zelichenko replied:
Yeah, that seems better. I wanted to be specific, but also thought it sounded awkward as is.

uttercm commented on civis/io/_tables.py:
```diff
@@ -1029,8 +1029,16 @@ def civis_file_to_table(file_id, database, table, client=None,
+                table_columns_have_sql = False
+                break
+
+    need_table_columns = (((not table_exists or existing_table_rows == 'drop')
```
I'm now wondering why we shouldn't throw an error when the table_columns are provided but don't have a `sql_type`, seems to me we might want to throw an error rather than ignore the provided ones?

izzy-zelichenko replied:
@uttercm ahh do you mean because it'll override all of them, even though some of them might be provide?

uttercm replied:
ya exactly, so they might be confused if the sql_types they did provide are different than the actual types we create since not all were provided

izzy-zelichenko replied:
That's a good point, but that would be a change in approach, so let me touch base with John on that.

jkulzick replied:
This makes sense. Just to be clear if no sql types are provided we will detect them but if some but not all are provided we will raise an error.

izzy-zelichenko replied:
Ah okay, I misunderstood the criteria in the first place. Thanks!

jkulzick commented on civis/tests/test_io.py:
```diff
@@ -398,6 +398,147 @@ def test_civis_file_to_table_table_doesnt_exist(
+    @pytest.mark.civis_file_to_table
+    @mock.patch('civis.io._tables._process_cleaning_results')
+    @mock.patch('civis.io._tables._run_cleaning')
+    def test_civis_file_to_table_table_exists_all_sql_types_missing(
```
I think we want `table_does_not_exist` for this test. We're testing that if all sql types are missing that we run cleaning? But if the table exists then we don't need to run cleaning so we probably want to change this to the table not existing or the `existing_table_rows = 'drop'`.

izzy-zelichenko replied:
Ah - I misread the condition statement. I was so confused.

izzy-zelichenko replied:
done

jkulzick commented on civis/tests/test_io.py:
```diff
@@ -398,6 +398,147 @@ def test_civis_file_to_table_table_doesnt_exist(
+            .id = mock_import_id
+        self.mock_client.get_database_id.return_value = 42
+        self.mock_client.default_credential = 713
+
```
Here we want to mock the api client response for `self.mock_client.get_table_id` based on whether or not the table exists. If the table exists, it should look like [this](https://github.com/civisanalytics/civis-python/blob/6d632e4e71888bdc63aba7655442bdc715c5f04d/civis/tests/test_io.py#L224). If the table does not exist then it should look like [this](https://github.com/civisanalytics/civis-python/blob/6d632e4e71888bdc63aba7655442bdc715c5f04d/civis/tests/test_io.py#L322).

izzy-zelichenko replied:
Yeah I see now, since I had the logic statement wrong I miswrote the tests

izzy-zelichenko replied:
done

jkulzick commented on civis/tests/test_io.py:
```diff
@@ -398,6 +398,147 @@ def test_civis_file_to_table_table_doesnt_exist(
+            True,  # headers
+            'gzip',  # compression
+            'comma',  # delimiter
+            table_columns  # table_columns
```
In this test, the `table_columns` provided as the parameter (to `civis.io.civis_file_to_table`) should be different than what is returned by `m_process_cleaning_results` or our cleaning job. This is because that job is detecting sql types since sql types were not provided and the table does not exist or is being dropped. So I would expect this `table_columns` to return valid `sql_type` entries.

izzy-zelichenko replied:
yeah I was wondering that! Looking at the existing tests it looked like they were passing in the same table_columns so I was confused about that, but I copied them. I'll take a closer look!

izzy-zelichenko replied:
done

izzy-zelichenko replied:
Yeah that makes sense!

jkulzick commented on civis/tests/test_io.py:
```diff
@@ -398,6 +398,147 @@ def test_civis_file_to_table_table_doesnt_exist(
+            'escaped': False,
+            'execution': 'immediate',
+            'loosen_types': False,
+            'table_columns': table_columns,
```
here we should expect the csv import job to be called with the `table_columns` returned by cleaning, not with the `table_columns` provided as the parameter to `civis_file_to_table`.

izzy-zelichenko replied:
Yeah, same confusion as above. Is it me or are these tests really hard to read...?

izzy-zelichenko replied:
done

jkulzick commented on civis/tests/test_io.py:
```diff
@@ -398,6 +398,147 @@ def test_civis_file_to_table_table_doesnt_exist(
+    ):
+        table = "scratch.api_client_test_fixture"
+        database = 'redshift-general'
+        mock_file_id = 1234
+        mock_cleaned_file_id = 1235
+        mock_import_id = 8675309
+
+        self.mock_client.imports.post_files_csv.return_value\
+            .id = mock_import_id
+        self.mock_client.get_database_id.return_value = 42
+        self.mock_client.default_credential = 713
+
+        table_columns = [{'name': 'a', 'sql_type': 'INT'},
+                         {'name': 'b', 'sql_type': ''}]
+        m_process_cleaning_results.return_value = (
+            [mock_cleaned_file_id],
+            True,  # headers
+            'gzip',  # compression
+            'comma',  # delimiter
+            table_columns  # table_columns
+        )
+        m_run_cleaning.return_value = [mock.sentinel.cleaning_future]
```
Review what we need to mock based on what gets called before we raise the `ValueError`. For example, we shouldn't hit `post_files_csv` or `m_process_cleaning_results` yet so we shouldn't need to mock those.

We will check if the table exists so we should be mocking the api client (based on the test name the table exists). See [here](https://github.com/civisanalytics/civis-python/blob/6d632e4e71888bdc63aba7655442bdc715c5f04d/civis/tests/test_io.py#L224).

izzy-zelichenko replied:
Ah yes, good catch. I meant to remove these.

izzy-zelichenko replied:
done

izzy-zelichenko commented on civis/io/_tables.py:
```diff
@@ -1027,10 +1027,23 @@ def civis_file_to_table(file_id, database, table, client=None,
+        for i in range(0, len(table_columns)-1):
+            thisSqlType = table_columns[i]['sql_type']
+            nextSqlType = table_columns[i+1]['sql_type']
+            if not thisSqlType == nextSqlType:
```
Yeah this might be confusing to read, but actually putting the `not` before a string evaluates it as a boolean. So what I am checking here is that if the next sql type exists but the previous one does not, or vice versa, then we throw an error. Because that means that there is a mix between existing and non-existing sql types. If the error isn't thrown then the original assignment of True or False remains, the former means they are all provided and the latter means none are provided. I'm pretty sure this logic achieves what we wants and I like that it exits the loop as early as necessary. But let me use a test just to be sure!

jkulzick replied:
I think in Python we can arrive at a much simpler approach that is easier to read. Python has a lot of powerful built-ins.

For example - `sql_type_cnt = sum(1 for col in table_columns if col.get('sql_type'))`

izzy-zelichenko replied:
Yeah that is simpler, I'll go with this approach.

izzy-zelichenko replied:
Done. That was easy!
I haven't used Python in a really long time, so it's hard for me to pull out these tools.

izzy-zelichenko commented on civis/io/_tables.py:
```diff
@@ -1027,10 +1027,23 @@ def civis_file_to_table(file_id, database, table, client=None,
+    if table_columns:
+        table_columns_valid = True if table_columns[0]['sql_type'] else False
+        for i in range(0, len(table_columns)-1):
+            thisSqlType = table_columns[i]['sql_type']
```
Yes I know this, it was just out of habit haha

izzy-zelichenko replied:
done

izzy-zelichenko commented on civis/io/_tables.py:
```diff
@@ -1027,10 +1027,23 @@ def civis_file_to_table(file_id, database, table, client=None,
     # Use Preprocess endpoint to get the table columns as needed
     # and perform necessary file cleaning
-    need_table_columns = ((not table_exists or existing_table_rows == 'drop')
+    need_table_columns = (((not table_exists or existing_table_rows == 'drop')
                           and table_columns is None)
+                          or not table_columns_valid)
```
Yes I see now! I misunderstood the logic at first.

izzy-zelichenko replied:
done

izzy-zelichenko commented on civis/io/_tables.py:
```diff
@@ -1027,10 +1027,23 @@ def civis_file_to_table(file_id, database, table, client=None,
+                error_message = 'Some table columns do not ' \
+                                'have a sql type provided'
+                raise ValueError(error_message)
+                break
```
done

izzy-zelichenko commented on civis/io/_tables.py:
```diff
@@ -1027,10 +1027,23 @@ def civis_file_to_table(file_id, database, table, client=None,
+            thisSqlType = table_columns[i]['sql_type']
+            nextSqlType = table_columns[i+1]['sql_type']
+            if not thisSqlType == nextSqlType:
+                error_message = 'Some table columns do not ' \
```
done

izzy-zelichenko commented on civis/io/_tables.py:
```diff
@@ -1040,7 +1052,7 @@ def civis_file_to_table(file_id, database, table, client=None,
     )
 
-    table_columns = table_columns or cleaned_table_columns
+    table_columns = cleaned_table_columns if need_table_columns else table_columns
```
This logic was flawed, since we may need to clean even when table_columns are provided, so needed to be updated.

izzy-zelichenko commented on civis/tests/test_io.py:
```diff
@@ -320,7 +320,8 @@ def test_civis_file_to_table_table_doesnt_exist(
         self.mock_client.get_table_id.side_effect = ValueError('no table')
-        mock_columns = [{'name': 'foo', 'sql_type': 'INTEGER'}]
+        mock_columns = [{'name': 'foo', 'sql_type': 'INTEGER'},
+                        {'name': 'bar', 'sql_type': 'VARCHAR(42)'}]
```
I added a second, different sql type to show that this test passes with current logic

izzy-zelichenko commented on civis/io/_tables.py:
```diff
@@ -1027,10 +1027,22 @@ def civis_file_to_table(file_id, database, table, client=None,
+        for i in range(0, len(table_columns)-1):
+            this_sql_type = table_columns[i]['sql_type']
+            next_sql_type = table_columns[i+1]['sql_type']
+            if (not this_sql_type) == (next_sql_type):
```
This was not actually working as expected, had to add parentheses to have them evaluate as needed.

uttercm replied:
See John's comments about simplifying this approach, this takes a bit of effort to understand what exactly we are looking for. I like the idea of using the counts of how many `sql_type` are provided.

izzy-zelichenko commented on civis/tests/test_io.py:
```diff
@@ -398,6 +398,242 @@ def test_civis_file_to_table_table_doesnt_exist(
+    @pytest.mark.civis_file_to_table
+    @mock.patch('civis.io._tables._process_cleaning_results')
+    @mock.patch('civis.io._tables._run_cleaning')
+    def test_civis_file_to_table_table_doesnt_exist_valid_table_columns_provided(
```
I added a test to show that when the table does not exist, but all sql_types are provided, that even if they are different, it does not throw an error and evaluates them correctly. No cleaning happens.

izzy-zelichenko commented:
We now have a test for every case when table_columns are provided

izzy-zelichenko commented on civis/io/_tables.py:
```diff
@@ -1027,10 +1027,22 @@ def civis_file_to_table(file_id, database, table, client=None,
     except ValueError:
         table_exists = False
 
+    table_columns_valid = True
```
got it, yeah agree!

izzy-zelichenko replied:
done. hows this?

izzy-zelichenko commented:
I removed the test I previously added because I realized it was nearly identical to the test `test_civis_file_to_table_table_doesnt_exist_provide_table_columns`, so I just updated that existing test to have two table_columns provided to be more thorough.

mheilman commented:
@izzy-zelichenko, would you mind making a followup PR to add an item to the changelog for this?

izzy-zelichenko commented:
@mheilman Sure thing.
